### ISSUE TITLE
com.microsoft.sqlserver.jdbc.KerbAuthentication

### ISSUE THREAD

**Issue by jangac-cyber** [microsoft/mssql-jdbc#2401]
## Question
Hello,

I am trying to connection from Apache NIFI to a SQL database using de MSSQL-JDBC driver version 12.6.1 and/or 12.2.0. In both cases I am receiving the following error: 

Could not initialize class **com.microsoft.sqlserver.jdbc.KerbAuthentication** 

This happens if I use the keytab credential services (KeytabCredentialsService) and/or using a username and password.

jdbc:sqlserver://<sqlserver>;database="<database>";integratedSecurity=true;trustServerCertificate=true;encrypt=true;authenticationScheme=JavaKerberos;

If I check usign kinit I can validate without any errors so both the username\password combination and the keytab are valid. 

full error message: **Encountered unexpected failure when attempting to perform verification: java.lang.NoClassDefFoundError: Could not initialize class com.microsoft.sqlserver.jdbc.KerbAuthentication**

**Comment by barryw-mssql** [comment]
Hi jangac-cyber,

Normally when we see an error like the one you are reporting, there is a problem with the classpath or something related to the class lookup. This error does not normally happen in a properly configured environment with the driver. Can you please provide more details of your environment so we may understand the contributing factors?

Thx,
Barry

**Comment by jangac-cyber** [comment]
Hi Jeffery-Wasty,

It is a RHEL 9.2 environment with Apache NIFI 2.0M2 and with JDBC library mssql-jdbc-12.6.1.jre11.jar in the /opt/nifi/lib 

Also, in the NIFI bootstrap conf file I have added the following three line

java.arg.18=-Djava.security.auth.login.config=/opt/nifi/conf/jaas.conf
java.arg.19=-Djava.security.krb5.conf=/opt/nifi/conf/krb5.ini
java.arg.20=-Djava.security.krb5.debug=true

The jaas conf (jaas.conf)  contains the following info:

SQLJDBCDriver {
  debug = true
  com.sun.security.auth.module.Krb5LoginModule required
  doNotPrompt=true
  useTicketCache=true;
};

In the log file of Apache NIFI the following java error in mentioned
`java.lang.NoClassDefFoundError: Could not initialize class com.microsoft.sqlserver.jdbc.KerbAuthentication
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.logon(SQLServerConnection.java:5352)
        at com.microsoft.sqlserver.jdbc.SQLServerConnection$LogonCommand.doExecute(SQLServerConnection.java:5334)
        at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7739)
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:4384)
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.connectHelper(SQLServerConnection.java:3823)
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.login(SQLServerConnection.java:3348)
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.connectInternal(SQLServerConnection.java:3179)
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.connect(SQLServerConnection.java:1953)
        at com.microsoft.sqlserver.jdbc.SQLServerDriver.connect(SQLServerDriver.java:1263)
        at org.apache.commons.dbcp2.DriverConnectionFactory.createConnection(DriverConnectionFactory.java:52)
        at org.apache.commons.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:414)
        at org.apache.commons.dbcp2.BasicDataSource.validateConnectionFactory(BasicDataSource.java:113)
        at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:629)
        at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:535)
        at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:711)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:714)
        at java.base/javax.security.auth.Subject.doAs(Subject.java:525)
        at org.apache.nifi.security.krb.AbstractKerberosUser.doAs(AbstractKerberosUser.java:186)
        at org.apache.nifi.security.krb.KerberosAction.execute(KerberosAction.java:74)
        at org.apache.nifi.dbcp.AbstractDBCPConnectionPool.getConnection(AbstractDBCPConnectionPool.java:230)
        at org.apache.nifi.dbcp.AbstractDBCPConnectionPool.verify(AbstractDBCPConnectionPool.java:89)
        at org.apache.nifi.controller.service.StandardControllerServiceNode.verifyConfiguration(StandardControllerServiceNode.java:531)
        at org.apache.nifi.web.dao.impl.StandardControllerServiceDAO.verifyConfiguration(StandardControllerServiceDAO.java:433)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:343)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:751)
        at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:751)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:703)
        at org.apache.nifi.web.StandardNiFiServiceFacade.performControllerServiceConfigVerification(StandardNiFiServiceFacade.java:2945)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:343)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:751)
        at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:751)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:703)
        at org.apache.nifi.web.api.ControllerServiceResource.lambda$performAsyncConfigVerification$28(ControllerServiceResource.java:1108)
        at org.apache.nifi.web.api.concurrent.AsyncRequestManager$2.run(AsyncRequestManager.java:119)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.SecurityException: java.io.IOException: Configuration Error:
        at java.base/sun.security.provider.ConfigFile$Spi.<init>(ConfigFile.java:139)
        at java.base/sun.security.provider.ConfigFile.<init>(ConfigFile.java:104)
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
        at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:132)
        at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:259)
        at java.base/java.lang.Class.newInstance(Class.java:755)
        at java.base/javax.security.auth.login.Configuration$2.run(Configuration.java:258)
        at java.base/javax.security.auth.login.Configuration$2.run(Configuration.java:249)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:571)
        at java.base/javax.security.auth.login.Configuration.getConfiguration(Configuration.java:248)
        at com.microsoft.sqlserver.jdbc.KerbAuthentication.<clinit>(KerbAuthentication.java:48)
`

**Comment by barryw-mssql** [comment]
Hi jangac-cyber,

Thanks for the information. From the call stack you provided, the error is not caused by the Kerberos settings in Apache NiFi but caused because the com.microsoft.sqlserver.jdbc.KerbAuthentication class cannot be found by Java or a subsequent class loader (i.e. Kerberos configuration is not a factor in this error). If you use a ZIP program to open up the driver JAR file you should be able to see that the KerbAuthentication class file is contained in the JAR. As the class is part of the JAR you need to ensure that Java or program class loader can find the driver JAR. Please confirm that the driver JAR is in the class path used by Apache NiFi (i.e. Apachi NiFi may use a different built-in classpath separate from the environment classpath variable).

Note: This Data Direct blog (https://www.progress.com/tutorials/jdbc/ingest-salesforce-data-incrementally-into-hive-using-apache-nifi) for installing a JDBC driver in Apache NiFi recommends putting the installed driver JAR in the classpath and then copying the driver JAR file(s) to <path>/nifi/lib. This means that the Data Direct driver is both in a public classpath and the NiFi classpath. This should hold true for the SQL Server JDBC driver.

Please let us know if this helps.

Thx,
Barry

**Comment by jangac-cyber** [comment]
Hi Barry,

Thank you for your support, but I finally figured out what was causing the KerbAuthentication error. 

The problem was in de jaas.conf file, the debug = true value was wrong 
wrong jaas.conf:

> SQLJDBCDriver {
> debug = true
> com.sun.security.auth.module.Krb5LoginModule required
> doNotPrompt=true
> useTicketCache=true;
> }; 

correct jaas.conf:

> SQLJDBCDriver {
> com.sun.security.auth.module.Krb5LoginModule required debug=true
> doNotPrompt=true
> useTicketCache=true;
> };

Using java -Djava.security.auth.login.config=jaas.conf JaasAcn.java file I managed to figure it out. The JaasAcn.java is taken from [Introduction to JAAS and Java GSS-API Tutorials] (https://docs.oracle.com/javase/8/docs/technotes/guides/security/jgss/tutorials/JaasAcn.java)](url)

I got the same error as mentioned in java call stack **java.io.IOException: Configuration Error: at** but with the line on which the error occured. Once I removed the **debug =true** it works and after using debug=true after _Krb5LoginModule required_ is still worked

Regards
Charl
